### PR TITLE
Change create organization flow

### PR DESCRIPTION
**Pull request by biswajit-hypto**
Problem:
We were currently receiving user/org details for create org after email verification. We will now change the procedure for creating an organisation such that we gather all org/user data before validating email.

Solution:
All org and root user details will be stored as metadata in the passcode table, which will be used to create org and root users after the user calls the create org api with the received passcode.

Changes:
Changed the Create Org Api to allow for the creation of organisations in two methods.
1. If a secret key is supplied, construct org using the data from the request body.
2. Alternatively, use metadata from the passcode table to create an org and a root user.

Refactored tests to create org using new flow

### CODE REVIEW

**Comment by biswajit-hypto** [comment]
Remaining tasks:

- [ ] Encrypting metadata as confidential info like password of root user is present there.

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt [hunk 1]
```diff
@@ -28,17 +33,27 @@ import org.koin.ktor.ext.inject
...
-        val passcodeStr = call.principal<ApiPrincipal>()?.tokenCredential?.value
-        val (organization, tokenResponse) = service.createOrganization(
-            request.name,
-            description = request.description ?: "",
-            rootUser = request.rootUser,
-            passcodeStr = passcodeStr
+        val secretKey = Constants.SECRET_PREFIX + appConfig.app.secretKey
+        val passcodeStr = call.principal<ApiPrincipal>()?.tokenCredential?.value!!
+
+        // If secretKey is provided, then fetch use request body else use metadata from passcode table
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt:45]: IAM should support different sign-up strategies.
1. User already entered admin details before verification.
2. User enters admin details after email verification.
3. Partial sign-ups where user enters credentials but email is not verified.

Logic here can be to
1. Check passcode table or sign-up request object and if the admin details are available, create org
2. If data is available in both the places or if the data is available in no place, throw 400.
3. For partial sign-ups support, we can give temporary token to play around in sandbox. (This can be added later)

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt [hunk 2]
```diff
@@ -28,17 +33,27 @@ import org.koin.ktor.ext.inject
...
+        val request = if (secretKey == passcodeStr)
+            call.receive<CreateOrganizationRequest>()
+        else {
+            val passcode = passcodeRepo.getValidPasscode(passcodeStr, VerifyEmailRequest.Purpose.signup)
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt:49]: If the request has an invalid passcodeStr then this method will throw exception. Is it handled properly by throwing bad request exception to customer ?

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt (hunk 1 above)
Comment (biswajit-hypto) [reply src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt]: Added support for 1 & 2

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt (hunk 2 above)
Comment (biswajit-hypto) [reply src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt]: All the validation regarding valid `passcode` is done in authentication level.

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt [hunk 3]
```diff
@@ -28,17 +33,33 @@ import org.koin.ktor.ext.inject
...
-            request.name,
-            description = request.description ?: "",
-            rootUser = request.rootUser,
-            passcodeStr = passcodeStr
+        val passcodeStr = call.principal<ApiPrincipal>()?.tokenCredential?.value!!
+
+        val apiRequest = call.receiveOrNull<CreateOrganizationRequest>()
+        val passcodeMetadata = passcodeRepo.getValidPasscode(passcodeStr, VerifyEmailRequest.Purpose.signup)?.metadata
+        if (passcodeMetadata != null && apiRequest != null) {
+            throw BadRequestException("Organization and root user details are already provided")
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt:46]: nit: Update the error message

"Admin user details are provided both during passcode creation time and sign-up request."

File: src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt [hunk 4]
```diff
@@ -28,17 +33,33 @@ import org.koin.ktor.ext.inject
...
+        }
+        if (passcodeMetadata == null && apiRequest == null) {
+            throw BadRequestException("Organization and root user details are missing")
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/apis/OrganizationApi.kt:49]: Admin user details are missing.

File: src/main/kotlin/com/hypto/iam/server/service/OrganizationsService.kt [hunk 5]
```diff
@@ -47,41 +44,25 @@ class OrganizationsServiceImpl : KoinComponent, OrganizationsService {
...
-    private val appConfig: AppConfig by inject()
 
     override suspend fun createOrganization(
-        name: String,
-        description: String,
-        rootUser: RootUser,
-        passcodeStr: String?
+        request: CreateOrganizationRequest
     ): Pair<Organization, TokenResponse> {
-        val secretKey = Constants.SECRET_PREFIX + appConfig.app.secretKey
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/service/OrganizationsService.kt:58]: Why is this removed ?

File: src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt [hunk 6]
```diff
@@ -35,21 +37,26 @@ class PasscodeServiceImpl : KoinComponent, PasscodeService {
...
-            purpose
-        )
+        val passcodeRecord = PasscodesRecord().apply {
+            this.id = idGenerator.passcodeId()
+            this.email = email
+            this.purpose = if (purpose == Purpose.signup) null else organizationId
+            this.validUntil = validUntil
+            this.purpose = purpose.toString()
+            this.createdAt = LocalDateTime.now()
+            // TODO: Store metadata in encrypted form as critical information like password is present during signup
```
Comment (hari-hypto) [inline src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt:56]: Can we encrypt as part of this PR itself ?

File: src/main/kotlin/com/hypto/iam/server/service/OrganizationsService.kt (hunk 5 above)
Comment (biswajit-hypto) [reply src/main/kotlin/com/hypto/iam/server/service/OrganizationsService.kt:58]: We will removing all passcodes related to that email and purpose regardless the api is called using the `secretKey`  or a valid `passcode`, so removing this variable as there is no need of it anywhere.

File: src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt (hunk 6 above)
Comment (biswajit-hypto) [reply src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt:56]: As of now we are using Elliptic Curve cryptography, which can only be used for signing the data and not encrypting it.
Can we do that in a different PR as we need to add and store new type of keys also.

Comment (biswajit-hypto) [reply src/main/kotlin/com/hypto/iam/server/service/PasscodeService.kt]: Added an AES encryption util to encrypt and decrypt metadata using the masterKey(privateKey) as the secret.

**Review by hari-hypto — APPROVED** [review]
